tensorflow/core/kernels/crop_resize_bilinear_core.h

```diff
@@ -0,0 +1,5651 @@
+                    _mm_slli_si128(v3, 2));
+}
+template <class T>
+void VectorLoader<T>::pack4_4b_3ch_(__m128i& v0, __m128i& v1, __m128i& v2,
+                                    __m128i& v3) {
+  __m128i and_mask = _mm_setr_epi32(-1, -1, -1, 0);
+  v0 = _mm_or_si128(_mm_and_si128(v0, and_mask), _mm_slli_si128(v1, 12));
+  and_mask = _mm_srli_si128(and_mask, 4);
+  v1 = _mm_or_si128(_mm_and_si128(_mm_srli_si128(v1, 4), and_mask),
+                    _mm_slli_si128(v2, 8));
+  and_mask = _mm_srli_si128(and_mask, 4);
+  v2 = _mm_or_si128(_mm_and_si128(_mm_srli_si128(v2, 8), and_mask),
+                    _mm_slli_si128(v3, 4));
+}
+
+template <>
+void VectorLoader<uint8>::pack_1ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                   __m128i& v3) {
+  pack4_1b_1ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<int8>::pack_1ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                  __m128i& v3) {
+  pack4_1b_1ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<uint16>::pack_1ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                    __m128i& v3) {
+  pack4_2b_1ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<int16>::pack_1ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                   __m128i& v3) {
+  pack4_2b_1ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<int32>::pack_1ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                   __m128i& v3) {
+  pack4_4b_1ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<Eigen::half>::pack_1ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                         __m128i& v3) {
+  pack4_2b_1ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<bfloat16>::pack_1ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                      __m128i& v3) {
+  pack4_2b_1ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<float>::pack_1ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                   __m128i& v3) {
+  pack4_4b_1ch_(v0, v1, v2, v3);
+}
+
+template <>
+void VectorLoader<uint8>::pack_2ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                   __m128i& v3) {
+  pack4_1b_2ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<int8>::pack_2ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                  __m128i& v3) {
+  pack4_1b_2ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<uint16>::pack_2ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                    __m128i& v3) {
+  pack4_2b_2ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<int16>::pack_2ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                   __m128i& v3) {
+  pack4_2b_2ch_(v0, v1, v2, v3);
+}
+template <>
+void VectorLoader<int32>::pack_2ch(__m128i& v0, __m128i& v1, __m128i& v2,
+                                   __m128i& v3) {
+  pack4_4b_2ch_(v0, v1, v2, v3)
```

Comment:
This sequence of instructions (VADD, VMUL, VSUB) would benefit from using fused multiply-add instruction (VFMA) instead of using individual VMUL, VADD. For e.g., the latency of VADD and VMUL are 3 and 5 cycles respectively (total 8 cycles) whereas VFMA is 5 cycles on Intel Haswell (AVX2) architecture. The compiler might actually fuse them into a FMA, but it would be better to explicitly write out FMA’s since this is intrinsics code that is targeted for AVX2 ISA.

Reply:
Thanks for that suggestion Vamsi, I have changed the AVX2 code to explicitly use FMA instructions whenever possible and will update this pull request tomorrow.